### DISCUSSION TITLE
Generate clusters_ids from two different linker objects in the same session

### DISCUSSION THREAD

**Discussion by ABJ66** [moj-analytical-services/splink#2232] [General]
Hi Robin and team,

Thanks for building this package it is super-useful.

Look forward to your comments on the question below. 

**Background:**

I am trying to refine a linkage approach to get more matching records using Splink. There are 3 key features in both the source datasets as follows: (1) name, (2) name_alias, (3) address. There are other features like city, state etc which I am not using while configuring Splink settings. Instead I have access to latitude and longitude information per unique_id of each data source. I am using this information as a blocking rule which considers only those candidate pairs which are roughly within a (haversine) distance of 2 miles.

Below is the model settings dictionary:
settings = {
    "link_type": "link_only",
    "blocking_rules_to_generate_predictions": block_rules_list,
    "comparisons": [
        ctl.name_comparison("name"), #cl.jaccard_at_thresholds("name")
        ctl.name_comparison("name_alias"),
        ctl.name_comparison("address")
    ],
    "retain_intermediate_calculation_columns" : True,
    "additional_columns_to_retain": ['city', 'lat_col', 'long_col', 'state', 'zip'], 
    "max_iterations": 50,
    "em_convergence": 0.005
    }

The above model settings generate around 2.5K links. 

However, there is a requirement to boost the links using deterministic_link() only even though it can return some false positives. Settings for such deterministic link is as below (blocking rule is when both the lat/long for records across sources is almost equal): 

settings_det = {
    "link_type": "link_only",
    "blocking_rules_to_generate_predictions": [""" round(l.lat_col, 6) = round(r.lat_col, 6) and round(l.long_col, 6) = round(r.long_col, 6) """],
    "comparisons": [],
    "retain_intermediate_calculation_columns": True,
    "additional_columns_to_retain": [c for c in df1.columns],
    "max_iterations":50,
    "em_convergence":0.005
}

linker_det = DuckDBLinker([df1, df2], settings_det,  input_table_aliases = ['source_a', 'source_b'])
xdf_det = linker_det.deterministic_link().as_pandas_dataframe(limit = None)
xdf_det.shape 

This new addition returns around 2.4K links. And after combining the resulting dataframe from both the approaches I get around 2.9K unique/non-overlapping links in total. 

**Question:**
The next task is to run the cluster_pairwise_predictions_at_threshold() on these combined results. I would like to know if there is a way to do so in Splink. It is a bit clumsy to do it outside Splink using say 'networkx'.

Earlier I tried few iterations to override the m & u probabilities from the 1st approach itself for features: name and name_alias & using either a combination of name + address or name_alias + address. but it did not give me much improvement. 

Thanks in advance.
Regards.

**Comment by RobinL** [comment]
Yeah, this is a reasonable thing to want to do that isn't easy.  But here's a runnable example that I believe works

```python
import duckdb

from splink.datasets import splink_datasets
from splink.duckdb.blocking_rule_library import block_on
from splink.duckdb.comparison_library import (
    exact_match,
    levenshtein_at_thresholds,
)
from splink.duckdb.linker import DuckDBLinker

df = splink_datasets.fake_1000

con = duckdb.connect()

settings = {
    "probability_two_random_records_match": 0.01,
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        block_on(["first_name"]),
        block_on(["surname"]),
    ],
    "comparisons": [
        levenshtein_at_thresholds("first_name", 2),
        exact_match("surname"),
        exact_match("dob"),
    ],
}


linker = DuckDBLinker(df, settings, connection=con)
predictions_1 = linker.predict()

settings = {
    "probability_two_random_records_match": 0.01,
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        block_on(["first_name"]),
        block_on(["surname"]),
    ],
    "comparisons": [
        exact_match("dob"),
        exact_match("email"),
        exact_match("city"),
    ],
}

linker2 = DuckDBLinker(df, settings, connection=con)
linker2._settings_obj._blocking_rules_to_generate_predictions = [
    block_on(["first_name", "surname", "dob"]),
]
predictions_2 = linker2.predict()
predictions_2.as_pandas_dataframe()
union_all_sql = f"""
select unique_id_l,  unique_id_r, match_probability
from
{predictions_1.physical_name}
UNION ALL
select unique_id_l,  unique_id_r,match_probability
from
{predictions_2.physical_name}


"""
unioned = con.sql(union_all_sql)

union_predict = linker2.register_table(
    unioned.df(), overwrite=True, table_name="my_table"
)
linker2.cluster_pairwise_predictions_at_threshold(
    union_predict, 0.5
).as_pandas_dataframe()

```

**Comment by ABJ66** [comment]
Hi Robin, 

Thanks for providing this solution. I am able to merge outputs of probabilistic and deterministic linkages with some modification of the above shared code. 

There is this final piece of adding manually identified links to be a part of the cluster_pairwise_predictions_at_threshold() which I am not able to add. These will typically be created after going through the list of unlinked ids in both the datasets which are to be linked. So there is a need to store the intermediate outputs e.g. trained models, settings etc, create the manually identified links, and add these back before running the cluster_pairwise_predictions_at_thresholds().

These are the steps I've followed: 

1. Create union_all_sql by combining variants of probabilistic linkages trained on different combinations of features (e.g. all, 2 or 3 and so on) + manually added linkages (using the same 'unique_id_l+unique_id_r' as the key in the blocking condition). A separate linker object is created for the manually added mappings. Details (settings dictionary etc) given in the next bullet.

2.
set_manual = {
    "link_type": "link_only",
    "blocking_rules_to_generate_predictions": [""" l.key = r.key """],
    "comparisons": [],
    "retain_intermediate_calculation_columns": True,
    "additional_columns_to_retain": 
        ['name', 'name_alias', 'address', 'city', ...],
  }

linker_blank = DuckDBLinker([df1, df2], settings_dict = {"link_type": "link_only"}, connection = con)

union_predict = linker_blank.register_table(
    unioned.df(), overwrite=True, table_name="my_table"
)

3. The above unioned_df contains some more number of rows as compared to w/o inclusion of manually identified linkages (done in a separate trial)

4. Lastly, calling cluster_pairwise_predictions_at_thresholds: 
`cluster_df0 = linker_blank.cluster_pairwise_predictions_at_threshold(union_predict, 0.5).as_pandas_dataframe(limit = None)`

However, I see that the number of rows in cluster_df0 is less by around 100 rows or so as compared to unioned_df.shape. Pls do suggest if there is a solution to match up the numbers amongst these two outputs. Since this new addition is a deterministic linkage, I do not think that the effect of using match_threshold_probability = 0.5 should have an impact on the clustering results in terms of number of rows.

Thank you for your support/guidance thus far.
